Add a new metric to measure apiserver request filter latency

tkashem opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Measure how much time a request spends in server filter(s):    
- A: the timestamp a particular filter starts executing for a request.
- B: the timestamp the particular filter has completed and the next handler in the chain has started executing.
    
We define filter latency as `B - A`. Introduce a new metric to track filter latency `apiserver_request_filter_duration_seconds`. This measures request filter latency distribution in seconds, for each filter type.

Enable filter latency tracking for the following filters:
- apf (priority and fairness)
- authentication
- authorization
- impersonation
- audit

Query to graph the filter latency by each request filter:
```
histogram_quantile(0.99, sum(rate(apiserver_request_filter_duration_seconds_bucket[5m])) by(filter,le))
```

The following screenshot shows filter latency in action on a dev cluster for `authentication`, `authorization`, `audit` and `impersonation`:
![image](https://user-images.githubusercontent.com/7385775/95923743-260a8580-0d84-11eb-9374-4198420363c8.png)

The following shows for priority and fairness
![image](https://user-images.githubusercontent.com/7385775/95923964-a16c3700-0d84-11eb-8f84-26d07907520c.png)

Both graphs are scaled to milliseconds.

```release-note
A new metric `apiserver_request_filter_duration_seconds` has been introduced that 
measures request filter latency in seconds.
```

tkashem commented:
@wojtek-t  @mm4tt I ran tests using the above server filters: https://github.com/tkashem/graceful/blob/master/priority-fairness/filter-latency/readme.md

wojtek-t commented:
/assign @MikeSpreitzer @yue9944882 @lavalamp

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -196,6 +196,18 @@ var (
 
+	requestFilterLatency = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Name: "apiserver_request_filter_duration",
```
What units are these in? Should that go in the name? Or at least the description?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -196,6 +196,18 @@ var (
+			Help: "How much time the request spends in apf",
+			Buckets: []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 15.0, 20.0, 25.0, 30.0,
+				35.0, 40.0, 45.0, 50.0, 60.0, 70.0, 80.0, 90.0, 100.0, 125.0, 150.0, 175.0, 200.0, 300.0,
+				400.0, 500.0, 600.0, 700.0, 800.0, 900.0, 1000.0},
```
The default settings will hold something for up to 15s.

How do you plan to represent requests that are held for the max time and then rejected?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -302,6 +315,25 @@ func UpdateInflightRequestMetrics(phase string, nonmutating, mutating int) {
+
+	ms := elapsed.Milliseconds()
+	if requestInfo.IsResourceRequest {
+		requestFilterLatency.WithLabelValues(name, cleanVerb(verb, req), requestInfo.APIGroup, requestInfo.APIVersion, requestInfo.Resource).Observe(float64(ms))
```
Oh it's ms? I think metrics are usually microseconds or seconds, at least I didn't expect this.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -196,6 +196,18 @@ var (
+		&compbasemetrics.HistogramOpts{
+			Name: "apiserver_request_filter_duration",
+			Help: "How much time the request spends in apf",
+			Buckets: []float64{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 15.0, 20.0, 25.0, 30.0,
```
My first thought: measure the .1-50ms region with high resolution, and then use exponential buckets up to 60s (4x the in-practice max with default flags).

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/filters/artificial_delayer.go:
```diff
@@ -0,0 +1,67 @@
+	"k8s.io/klog/v2"
+)
+
+// WithArtificialDelayAdder adds artificial delay to the .
```
cliffhanger

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go:
```diff
@@ -0,0 +1,90 @@
+	return info, ok
+}
+
+func DecorateFilter(chain http.Handler, name string, longRunningRequestCheck apirequest.LongRunningRequestCheck, decorate func(h http.Handler) http.Handler) http.Handler {
```
I don't understand this file at all, why is it so complicated? Is this for testing?

lavalamp replied:
I guess this is for timestamping before/after? I wouldn't use the word "decorate" for that, it's too general.

I'm not sure about this approach in general, P&F might already know the duration. I guess I'd expect to update this metric when P&F makes a decision about the request.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/filters/filter_decorator.go:
```diff
@@ -0,0 +1,90 @@
+
+const filterStartTimestampKey filterStartTimestampKeyType = iota
+
+func WithFiletrStartTimestamp(parent context.Context, startTimestamp time.Time) context.Context {
```
speeelling :)

tkashem commented:
@lavalamp I apologize, I should have made this clear. this is only for running a test. None of it is intended to be merged.

wojtek-t commented:
> @lavalamp I apologize, I should have made this clear. this is only for running a test. None of it is intended to be merged.

Ohh - I thought you wanted to merge that, it's my fault then.
[BTW - I think having those kind of metrics would actually be super useful in production too. So I think we should try to have that somehow in if possible.]

lavalamp commented:
Yeah I like this metric!

tkashem commented:
> [BTW - I think having those kind of metrics would actually be super useful in production too. So I think we should try to have that somehow in if possible.]

Hi @wojtek-t, that sounds good to me. I will refactor, make it more readable and then open it for review.

tkashem committed:
Measure how much time a request spends in server filter(s):

- A: the timestamp a particular filter starts executing for a request.
- B: the timestamp the particular filter has completed and the next handler
     in the chain has started executing.

We define filter latency as `B - A`. Introduce a new metric to track filter latency:
"apiserver_request_filter_duration_seconds"
This measures request filter latency distribution in seconds, for each filter type.

tkashem committed:
Enable filter latency tracking for request filters

Enable filter latency tracking for the following request filters:
- apf (priority and fairness)
- authentication
- authorization
- impersonation
- audit

tkashem commented:
> [BTW - I think having those kind of metrics would actually be super useful in production too. So I think we should try to have that somehow in if possible.]

Hi @lavalamp @wojtek-t I have enabled filter latency for a few filters, please review when you have time.

tkashem commented:
/retest

lavalamp commented:
/assign @logicalhan

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -196,6 +196,17 @@ var (
+		&compbasemetrics.HistogramOpts{
+			Name: "apiserver_request_filter_duration_seconds",
+			Help: "Request filter latency distribution in seconds, for each filter type",
+			Buckets: []float64{0.0, 0.0001, 0.0002, 0.0004, 0.0008, 0.001, 0.002, 0.004, 0.008, 0.01, 0.015, 0.02, 0.025,
```
overall LGTM, except for one nit: 30+ buckets looks too many to me.. isn't 10-ish buckets here just fine?

wojtek-t replied:
agree - that seems like an overkill to me.
All the filters should actually be pretty fast, so having such a bit granularity for larger values seems a bit useless. Especially for values above a second.

tkashem replied:
I set it to `[]float64{0.0, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.50, 1.0, 5.0}`. Let me know if this looks good.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:
```diff
@@ -42,6 +43,14 @@ func WithAudit(handler http.Handler, sink audit.Sink, policy policy.Checker, lon
 		return handler
 	}
+
+	// We want to track how much time a request spends in audit filter.
```
I initially though that having this kind of wrapper is overkill and preferred that to have more custom to the P&F.

But after thinking more about that and seeing this PR, I actually like the genericity of it and the fact we can use it for other filters too.
@lavalamp - does that seem reasonable to you too?

tkashem replied:
Hi @lavalamp @wojtek-t, I have enabled it for `priority and fairness`, `authentication` `authorization`, `impersonation` and `audit` filters for now.

sttts replied:
shouldn't we have this in the default handler chain builder func instead? The fan out would be smaller than adding it in all filters.

tkashem replied:
@sttts @MikeSpreitzer I have moved the tracking inside the `DefaultBuildHandlerChain` function now.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:
```diff
@@ -0,0 +1,92 @@
+type requestFilterRecord struct {
+	name               string
+	startedTimestamp   time.Time
+	completedTimestamp time.Time
```
Do we really need completedTImestamp?
It's an additional time to keep in memory, so maybe having name and startedTime is enough?

tkashem replied:
Yes, sounds good to me. I removed `completedTimestamp`

tkashem commented:
@MikeSpreitzer  https://github.com/tkashem/graceful/blob/master/priority-fairness/filter-latency/readme.md

tkashem commented:
/assign @deads2k @sttts

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -196,6 +196,16 @@ var (
+		&compbasemetrics.HistogramOpts{
+			Name:           "apiserver_request_filter_duration_seconds",
+			Help:           "Request filter latency distribution in seconds, for each filter type",
+			Buckets:        []float64{0.0, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.50, 1.0, 5.0},
```
Roughly exponential progression: good.
Note: the square root of 10 is about 3, not 5 or 2.

MikeSpreitzer replied:
How about
```suggestion
			Buckets:        []float64{0.0, 0.0003, 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1.0, 5.0},
```

tkashem replied:
sounds good, I made the change.

MikeSpreitzer reviewed (commented):
A few overall thoughts:

I agree with https://github.com/kubernetes/kubernetes/pull/95207#discussion_r510824094

Given that `context.WithValue(..)` always pushes an association, this builds up a long context chain.  That's a bit of an anti-pattern.  How about side-effecting when possible instead of pushing a new association?

I see this is measuring the overhead on the way in but not on the way out.  Would it be better to do both?

For the API Priority and Fairness filter, the current measurement covers two rather distinct sorts of time periods: the overhead of the filter, and the time spent waiting in a queue (in the cases where that happens).  It would be good to be able to observe the two separately.  Fortunately we already have `apiserver_flowcontrol_request_wait_duration_seconds` for the latter.  If all we care about is the average then we can subtract the average wait duration from the average total to get the average filter overhead.  If we want quantiles then this is not possible, rather a more direct observation of the filter overhead is needed.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -669,17 +670,39 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
+	handler = filterlatency.TrackCompleted(handler)
 	handler = genericapifilters.WithAuthentication(handler, c.Authentication.Authenticator, failedHandler, c.Authentication.APIAudiences)
+	handler = filterlatency.TrackStarted(handler, "authentication")
+
```
:+1: – a little more verbose here, but nicely contained where we define the chain.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -669,17 +670,39 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
+		// We want to track how much time a request spends in PriorityAndFairness filter.
+		handler = filterlatency.TrackCompleted(handler)
 		handler = genericfilters.WithPriorityAndFairness(handler, c.LongRunningFunc, c.FlowControl)
+		handler = filterlatency.TrackStarted(handler, "apf")
```
apf? priorityandfairness ?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -669,17 +670,39 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 
 func DefaultBuildHandlerChain(apiHandler http.Handler, c *Config) http.Handler {
-	handler := genericapifilters.WithAuthorization(apiHandler, c.Authorization.Authorizer, c.Serializer)
+	// We want to track how much time a request spends in authorization filter.
```
do we really need these comments?

sttts commented:
/approve
Lgtm – giving others time to review 

@logicalhan @wojtek-t @deads2k

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -669,17 +670,34 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
+
+	handler = filterlatency.TrackCompleted(handler)
 	handler = genericapifilters.WithImpersonation(handler, c.Authorization.Authorizer, c.Serializer)
+	handler = filterlatency.TrackStarted(handler, "impersonation")
```
can we factor this a little tighter?  Perhaps an `Add` method that does the wrapping or a filter around just the WithImpersonation call?

sttts replied:
note that we need both: start and complete.

tkashem replied:
@deads2k I initially had this pattern but it looked less readable imo. I had to use a function like this to get it into one function:
```
func Tracked(handler http.Handler, name string, tracked func(h http.Handler) http.Handler) http.Handler
```

MikeSpreitzer replied:
Yeah, either way would be OK with me.

wojtek-t commented:
> Lgtm – giving others time to review
> @logicalhan @wojtek-t @deads2k

LGTM - I'm leaving tagging it to others though.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:
```diff
@@ -0,0 +1,90 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		fr := &requestFilterRecord{
+			name:             name,
+			startedTimestamp: clock.Now(),
+		}
+		r = r.WithContext(withRequestFilterRecord(ctx, fr))
```
Building up a long chained context is an anti-pattern.  How about the following instead?

```suggestion
		if fr := requestFilterRecordFrom(ctx); fr != nil {
			fr.name = name
			fr.startedTimestamp = clock.Now()
		} else {
			fr := &requestFilterRecord{
				name:             name,
				startedTimestamp: clock.Now(),
			}
			r = r.WithContext(withRequestFilterRecord(ctx, fr))
		}
```

tkashem replied:
@MikeSpreitzer yeah, that sounds like a good idea! I made the change and added a test to verify.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency_test.go:
```diff
@@ -0,0 +1,177 @@
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+)
+
+func TestTrackStartedWithNoFilterRecordInContext(t *testing.T) {
```
I think you got the names of this and the following test swapped.

tkashem replied:
Fixed, thanks for catching it!

MikeSpreitzer reviewed (approved):
/LGTM

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:
```diff
@@ -196,6 +196,16 @@ var (
+		&compbasemetrics.HistogramOpts{
+			Name:           "apiserver_request_filter_duration_seconds",
+			Help:           "Request filter latency distribution in seconds, for each filter type",
+			Buckets:        []float64{0.0, 0.0003, 0.001, 0.003, 0.01, 0.03, 0.1, 0.3, 1.0, 5.0},
```
Having a 0.0 bucket doesn't make much sense - we should switch that to 0.0001 or sth like that.
@tkashem
